Support interserver credential rotation

johnskopis opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support dynamic interserver credentials

Detailed description / Documentation draft:
No changes are required to enable dynamic interserver credentials. 

When credentials are not configured the InterserverIOHandler will accept any credential as a valid credential. This is mostly compatible with previous versions but the behavior has changed. Previous versions of ClickHouse would disallow connections from clients that had configured authentication.

To enable authentication when it's not enabled specify `interserver_http_credentials.allow_empty` to permit unauthenticated clients.

To support rotating of server credentials, specify the previously used credentials under `interserver_http_credentials.users`. Including multiple credentials allows for clients to use either the current or previous credentials.

Here are some examples of seamlessly migrating to authenticated clients, and changing the credentials.

Assuming there are no credentials to start:

    <interserver_http_credentials />

Add a configuration to send authentication headers in client requests, while still accepting clients not using authentication:

    <interserver_http_credentials>
        <user>admin</user>
        <password>222</password>
        <allow_empty>true</allow_empty>
    </interserver_http_credentials>

Once all clients are authenticated remove the `interserver_http_credentials.allow_empty` to require all clients authenticate using `admin:222`.

To rotate credentials add a `interserver_http_credentials.users.{user}` to send the new credential while still accepting the old credential.

    <interserver_http_credentials>
        <user>admin</user>
        <password>333</password>
        <users>
            <admin>222</admin>
        </users>
    </interserver_http_credentials>

Once all clients are sending the new credential remove `interserver_http_credentials.users.{user}`
...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.


Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

Commit:
Support interserver credential rotation

Restarting a server instance to change the interserver password results
in many replicas being out of sync until all clusters are using the new
credential.

This commit adds dynamic credential loading for both the client
(Replicated* tables) and server (InterserverIOHTTPHandler).

This commit also adds the ability to rotate credentials, i.e. accept more
than one credential during a credential change.

state0 (no auth):

    <interserver_http_credentials />

state1 (auth+allow_empty migration):

    <interserver_http_credentials>
        <user>admin</user>
        <password>222</password>
        <allow_empty>true</allow_empty>
    </interserver_http_credentials>

state2 (auth+new admin password migration):

    <interserver_http_credentials>
        <user>admin</user>
        <password>333</password>
        <users>
            <admin>222</admin>
        </users>
    </interserver_http_credentials>

johnskopis commented:
fixes #14106

johnskopis commented:
I'm not sure if there are any concerns using shared_ptr from InterserverIOHandler. I noticed there's not many shared pointers used and am wondering if it's frowned upon in this codebase?

I was thinking it would be better to "bind" the credential store to a handler but this would require a bit of effort (handler manager singleton to swap the handler instance when the credentials change). I think shared_ptr here is simpler and requires less resources than copying the credentials for every request.

johnskopis commented:
Tests are green!
Should I move the code somewhere other than Interpreters? cc @filimonov

azat commented on programs/server/Server.cpp:
```diff
@@ -483,16 +484,11 @@ int Server::main(const std::vector<std::string> & /*args*/)
+    global_context->updateInterserverCredentials(config());
+    if (global_context->getInterserverCredential() == nullptr)
+        throw Exception("Error initializing interserver credentials",
+            ErrorCodes::NO_ELEMENTS_IN_CONFIG);
```
How is this possible?

johnskopis replied:
Correct, the code is unreachable. It's there in case it changes in the future

azat replied:
Same here, I'm suggesting removing it

akuzm replied:
We normally use a C++ reference, not pointer, for things that can't be nullptr. Didn't read the code, but maybe it's suitable here as well?

If it *can* be nullptr in some cases, but shouldn't be nullptr here -- a suitable exception code would be `LOGICAL_ERROR`. It is meant for conditions that should never occur and signify a program error.

johnskopis replied:
Ok, so I should update the callee to throw (LOGICAL_ERROR)  in case of an error and catch the reception in Server.cpp?

azat replied:
>We normally use a C++ reference, not pointer, for things that can't be nullptr. Didn't read the code, but maybe it's suitable here as well?

There is shared_ptr, indeed strange to see check against nullptr

But I actually suggesting remove this piece of code anyway.

johnskopis replied:
Server creates credentials when sever starts -- or throws and the server tattoos to start
When config is reloaded the server should never reset credentials with a nullptr. Existing code will never trigger the branch but future code might. Just trying to be safe here but since it does nothing it makes sense to remove

akuzm replied:
> Ok, so I should update the callee to throw (LOGICAL_ERROR) in case of an error and catch the reception in Server.cpp?

First, determine when it is correct for `getInterserverCredential()` to return nullptr. If it is never correct, it should return a reference, not a pointer.

If it can return nullptr, the next question is whether it shares ownership of interserver credentials with the caller. If not, returning a raw pointer is more suitable.

The next question is whether you want to meaningfully handle the nullptr here, or is it always a program error. If it is a program error, you don't really have to add a special check for nullptr, the server is going to fail with a segfault. We don't consistently check for null pointers everywhere, so this strategy is almost acceptable. If you prefer an exception, it should have the code LOGICAL_ERROR. You don't have to add any special handling for this exception (especially for this particular exception type), and can rely on the default handling, unless there is a reason not to do so (e.g. no default handling in this place of code).

johnskopis replied:
`getInterserverCredential` exists for InterserverIOHandler mainly. 

My expectation is that credentials should always point to something.

When the server is misconfigured it should fail to start. When the server is misconfigured on reload it should not update the credential (auth bypass). 

With an exception there's a side effect that causes config reload to fail (after updateCredentials). I can catch this in update credentials and skip the update though.

The reason is a shared ptr is to avoid copying or synchronization in InterserverIOHandler. I eluded to this in an earlier comment.

azat commented on src/Interpreters/Context.cpp:
```diff
@@ -1457,6 +1458,46 @@ bool Context::hasZooKeeper() const
+    std::lock_guard lock(shared->interserver_io_credentials_mutex);
+
+    if (credentials == nullptr)
+        return;
```
Looks superfluous

johnskopis replied:
Again, if it changes in the future

azat replied:
I'm suggesting remove it, did not see such checks things as a common thing

azat commented on src/Interpreters/InterserverCredentials.h:
```diff
@@ -0,0 +1,94 @@
+///            <admin>111</admin>
+///        </users>
+///    </interserver_http_credentials>
+class SimpleInterserverCredentials : public virtual BaseInterserverCredentials
```
Maybe smth like `ConfigInterserverCredentials` will be better?

johnskopis replied:
👍

azat commented on src/Interpreters/Context.h:
```diff
@@ -404,7 +405,9 @@ class Context
 
     /// Credentials which server will use to communicate with others
-    void setInterserverCredentials(const String & user, const String & password);
+    void setInterserverCredentials(std::shared_ptr<BaseInterserverCredentials> credentials);
```
Looks like should be private

johnskopis replied:
Private for context, sure.

azat commented on src/Interpreters/InterserverCredentials.cpp:
```diff
@@ -0,0 +1,69 @@
+        }
+    }
+    if (!valid)
+        return {"Incorrect user or password in HTTP Basic authentification", false};
```
why not via throw? Ok I see `InterserverIOHTTPHandler::checkAuthentication` is done in that way (but maybe it should be updated? (to ensure that the result will be 100% verified)

johnskopis replied:
Can throw if you want.StatusOr is probably less overhead (it's called from http handler). It's there a reason why the handler should throw?

azat replied:
This will avoid further issues if someone will forget to check the return value (and since this is authorization this looks pretty critical)

P.S. if you will change this, you will need to guard the upper caller to wrap the call into the try/catch (this can be done separately I guess)

johnskopis replied:
IsValid shouldl throw and checkAuth will catch

azat commented on src/Interpreters/InterserverCredentials.h:
```diff
@@ -0,0 +1,94 @@
+class SimpleInterserverCredentials : public virtual BaseInterserverCredentials
+{
+public:
+    using Store = std::vector<std::pair<std::string, std::string>>;
```
Maybe `std::unordered_map` will be better? (Will not require iteration through all of the users, but not sure that it worth it, since it should have too much elements, but at least it will make interface cleaner)

johnskopis replied:
Can do, but I think it's the same?

azat replied:
The usage will became simpler, right now it iterates over vector (instead of using `find()`)

johnskopis replied:
Use case was only to have 1 extra user in store (previous credentials). Using a map would allow the admin to use a credential per server. Sgtm

johnskopis replied:
Actually a map doesn't work because more than one password (old, new) should work for the same user

azat replied:
Ok, then it can be multimap, or at least just replace the looping over store with the `std::find`

johnskopis replied:
I replaced with a smap<pair<string, string>, bool>. I'm sure it's not what you had in mind but it's using std::find :)

azat commented on src/Interpreters/InterserverCredentials.h:
```diff
@@ -0,0 +1,94 @@
+#pragma once
```
>Should I move the code somewhere other than Interpreters?

Why not put it near the `InterserverIOHTTPHandler.cpp`?

johnskopis replied:
It's used from shared (replicated* and also interserver handler). Closer to the handler makes more sense for handler but not client. what do you think about storing client + server closer to usage? (Moving client creds back to shares)

azat replied:
Not sure that I got this, but now I noticed that it is used in the Interpreters too, so looks ok to me as-is

azat commented:
> I think shared_ptr here is simpler and requires less resources than copying the credentials for every request.

In the same manner clusters update works, so I guess it is fine

azat commented on tests/integration/test_replication_credentials/test.py:
```diff
@@ -127,6 +127,54 @@ def test_credentials_and_no_credentials(credentials_and_no_credentials_cluster):
     time.sleep(1)
 
-    assert node7.query("SELECT id FROM test_table order by id") == '111\n'
+    assert node7.query("SELECT id FROM test_table order by id") == '111\n222\n'
```
But why it now contains the newly inserted record? AFAICS it still should fail due to authentication error

johnskopis replied:
Node7 has no credentials configured so it accepts everything. This is useful to migrate from no credentials to configured credentials (unauthenticated server should allow authenticated client until server is reconfigured).. configuring a credential instead in null would retain existing behavior

azat replied:
But it is node8 who do not have credentials, am I missing something?

johnskopis replied:
Node7 can replicate from node8; node8 can't replicate from node7. Null credentials allow any connection. This change in behavior is documented in message.

johnskopis replied:
Previously if the client sent credentials but auth want configured on the server it would reject. With this change ba server w/o auth configured accepts any credential

johnskopis commented:
I *think* I have addressed all the feedback. Please take a look.
Also, I'm not sure what to do about:
flake: Just delete it?
broken clang11 build -- it doesn't like the naming convention what should I do instead?

nvartolomei commented on src/Interpreters/InterserverCredentials.cpp:
```diff
@@ -0,0 +1,61 @@
+    return store_;
+}
+
+bool ConfigInterserverCredentials::isValidUser(const std::pair<std::string, std::string> check)
```
nit: s/check/userAndPass? same in .h file (the argument names should match to make clang11 build happy)

nvartolomei commented on src/Interpreters/InterserverCredentials.cpp:
```diff
@@ -0,0 +1,61 @@
+    if (user.empty())
+        throw Exception("Configuration parameter interserver_http_credentials user can't be empty", ErrorCodes::NO_ELEMENTS_IN_CONFIG);
+
+    auto store_ = makeCredentialStore(user, password, config, root_tag);
```
just remove the `_`? for clang11 same on line 33

alesapin committed:
Merge branch 'master' into johnskopis/dynamic-interserver-creds-v20

alesapin committed:
Buildable code

alesapin committed:
Some fixes

alesapin committed:
fix style

alesapin committed:
Fix duplicated headers

alexey-milovidov commented:
```
<interserver_http_credentials>
    <user>admin</user>
    <password>333</password>
    <users>
        <admin>222</admin>
    </users>
</interserver_http_credentials>
```

This example has inconsistency: the user name is
- in the content of `<user>` tag;
- in tag name: `<admin>`.

Writing user names in tag name has limitations.

The more consistent way will be:
```
<interserver_http_credentials>
    <user>admin</user>
    <password>333</password>
    <old>
        <user>admin</user>
        <password>222</password>
    </old>
    <old>
        ...
</interserver_http_credentials>
```

Or (more simple, maybe acceptable)?
```
<interserver_http_credentials>
    <user>admin</user>
    <password>333</password>
    <password>222</password>
</interserver_http_credentials>
```

alesapin committed:
Merge branch 'master' into johnskopis/dynamic-interserver-creds-v20

alesapin committed:
Small improvements

alesapin committed:
Better interface

alesapin committed:
More clear messages

alesapin committed:
Fix PVS

alesapin commented:
No related failures. Let's merge.

alexey-milovidov commented:
@alesapin It would be better to also put explanation and example in config.xml.

johnskopis commented on src/Interpreters/Context.cpp:
```diff
@@ -1735,40 +1734,15 @@ bool Context::hasAuxiliaryZooKeeper(const String & name) const
-{
-    std::lock_guard lock(shared->interserver_io_credentials_mutex);
-    shared->interserver_io_credentials = credentials;
+    return shared->interserver_io_credentials.get();
```
shouldn't this return the smart pointer?

johnskopis replied:
nevermind, I see a MultiVersion now :)

johnskopis commented:
thanks for finally merging this one!

gyuton commented:
Internal documentation ticket: DOCSUP-15663.
